### ISSUE TITLE
Doc Site: Markdown POC

### ISSUE THREAD

**Issue by samrichca** [overture-stack/roadmap#87]
See my and Rob's investigation here:

https://app.zenhub.com/workspaces/argo-mega-board-all-repos-5c8c8637777e5f767e486217/issues/overture-stack/roadmap/72

I want to resolve the question of whether distributed documentation will work for us, and establish conventions for writing documentation in Markdown.

We know we can get Markdown files from other repos at build time but - does the rest of it work?

- How do we organize/order documentation pages so we can navigate between them?
  - **YAML files for now.**
- Can we have subsections? How? Folders? i.e. /documentation/song/installation/mac
  - **Yes**
- Do all folders need an index.md or can we stub that in?
  - **Folders need an index but we can stub it in: https://joshuatz.com/posts/2019/gatsby-automatic-directory-listing-pages-beginners-attempt/**
- Will Algolia search work?
  - **Algolia search looks at Gatsby's GraphQL endpoint. How data gets there is up to us.**
- What happens in local development? (with remote docs)
  - **You should add an env variable to make fetching docs optional. Otherwise it fetches every time Gatsby clears it cache. So, only fetch remote docs if you need them.**
- How can we combine local & remote MD files?
  - **Yes, they're not treated differently at all. They both get imported as "Markdown nodes".**
- What happens with images, Youtube embeds, and other media?
  - **This all works the same with local & remote docs.**
- Any Netlify issues?
  - **I got rate-limited by GitHub. Need to save GitHub auth secrets in Netlify profile. Otherwise no.**
- Any other technical limitations?
  - **Gatsby fails with untraceable errors, and all docs have to be fetched on every build. What if there are problems in Song's repo and that breaks the website build, how do we track that down?**
- What is needed for MVP and what could be an enhancement?
  - **Investigating a Gatsby approach for this was worthwhile, i.e. I know more about Gatsby now, but we should pursue a more sophisticated solution if any.**
- What are the technical conventions we need to use for Markdown files?
  - Do they need front matter? i.e. title, date
    - **No date needed. If title isn't provided, the filename will be used. Folders need an index.md unless we decide to generate stubs.**
- Can we disable this with the SHOW_DOCS environment variable? (i.e. feature flag it)
  - **Yes. Pages aren't created if the flag is false.**

---

## scope

What I anticipate doing for this ticket is
- [x] Get stuff working. It can be ugly/unstyled for now.
- [x] Add local pages
- [x] Add a sitemap JSON for the local pages.
- [x] Add prev/next buttons.
- [x] Add table of contents: section contents (pages - based on sitemap JSON) and page contents (headings - this is generated by the graphQL request)

### removed from scope for this ticket

- Code snippets, tables, warning/note boxes. They work but they have no styling and just look like lorem ipsum. **Will do later**
- Add remote pages **(decided not to do this)**
  - Add these pages to the navigation.
  - Try single pages, glob patterns, folders, etc.
  - Try develop & production builds.
  - Try deploying on Netlify.
  - Rob and I investigated how to do this and he made a small POC already.

**Comment by samrichca** [comment]
pre-long weekend update:

- I have a `hello world` markdown page
- index.md/mdx files become the root page of their directory
- support for files without titles
- support for feature flag

**Comment by b-f-chan** [comment]
Decision - Max 4 levels of pages for starters?

JetBrains example: https://www.jetbrains.com/help/idea/discover-intellij-idea.html

**Comment by samrichca** [comment]
based on sprint planning today:

this can all be changed but it's what i'm doing for now based on examples provided.

- file folders: one level: /documentation/song/
- URLs: 2 levels: /documentation/song/installation/ (file path: /documentation/song/installation.md)
- left sidebar navigation: up to 4 levels, described in a YAML file
- sidebar nav headings can link to a page, or just open/close a menu.
- 1 file to 1 page, not concatenating files.
- prev/next buttons will attempt to find the prev/next nav items in the same section (song is a section) that are links.
  - more technical explanation: flatten nav object, find prev/next sibling with file property.

**Comment by samrichca** [comment]
(I think ZenHub ate my original comment?)

I'm parking the distributed documentation idea, but we could re-visit it as a future enhancement.

Goal: 
- Define project repos to get docs from, and specify a release tag.
- Re-fetch only when the release tag changes, either by keeping the docs in Git or by using differential builds (only fetching/building the files that changed).

Result:
- Docs are re-fetched/built very often. They're saved in Gatsby's cache which is cleared often in dev, and in every prod build.

Issues:
- GitHub has rate limits and we need to add auth (secrets are supported by Gatsby & Netlify)
- If a fetch/build fails, the website won't build, and the error messaging is unclear. We'd have to debug across several repos. 
- We have to do test-runs locally.
- Gatsby doesn't support differential builds for Markdown sites.
- Can't save docs files to the website's git repo unless we pull them from the default branch.
- **TL;DR fetching has to happen frequently, which will introduce untraceable errors, and doesn't reduce friction for contributors.**

Solutions:
- **Copy/paste docs from the project repos when we want to update the docs on the website. <<< this is what I'm doing for now**
- Explore more sophisticated CI/CD solutions, move the website off of Netlify, etc (out of scope for this project).

**Comment by samrichca** [comment]
##  Discussion point:

Here's how the documentation works for now, which isn't final, but allows the user to navigate between pages so I can move forward with other tickets.

- External documentation has been removed, but it's in the commit history in case we want to revisit it.
- URLs will look like /documentation/song/the-page-name
- Folder/URL structure is flat for now. If we want to nest it, we should make sure there are index files for each level.
- We can change this later. Essentially, Gatsby puts everything into GraphQL, so we can do anything we want, as long as it's consistent/programmable. The folders can look different from the URLs if we want.

![image.png](https://images.zenhubusercontent.com/5d2e194240c06427886a21ee/73ef2914-7e75-4122-a0bc-15e537faee89)

Note: We still have nesting in the navigation, just not in the folder or URL structure right now.

![image.png](https://images.zenhubusercontent.com/5d2e194240c06427886a21ee/7c812c47-aa3d-429a-b94c-be3c02c8be9d)

**Comment by samrichca** [comment]
## Testing:

![image.png](https://images.zenhubusercontent.com/5d2e194240c06427886a21ee/9ca69b58-b5d2-4492-92bc-6e6e08f24508)

And then I deleted 2 of the sections but I didn't want to remake all of these images. :)

![image.png](https://images.zenhubusercontent.com/5d2e194240c06427886a21ee/6724f0dd-f811-4129-b513-51008e27d06c)

![image.png](https://images.zenhubusercontent.com/5d2e194240c06427886a21ee/7a7d774d-4242-4fd3-8c8f-88b417ea7104)

![image.png](https://images.zenhubusercontent.com/5d2e194240c06427886a21ee/f3aa2c39-ae1c-44c2-8f00-c9492f454207)

![image.png](https://images.zenhubusercontent.com/5d2e194240c06427886a21ee/487477f7-e1f9-49de-9519-cf5da96c1a68)

**Comment by samrichca** [comment]
WITH DOCS
https://5f9877737040440e8dc0cb7e--overture.netlify.app/documentation/

NO DOCS
https://5f9878425f98eb0d7af89240--overture.netlify.app/documentation/ - 404 page

**Comment by samrichca** [comment]
Updated netlify with megamenu

WITH DOCS
https://5f997e4abfce39b83f78d6d8--overture.netlify.app/documentation/

NO DOCS 
https://5f997ebd26e9f8a874aba47a--overture.netlify.app/documentation/

**Comment by b-f-chan** [comment]
Testing will carryover to Sprint 9 (cc: @rosibaj please have a look as well if you have time)

**Comment by kcullion** [comment]
@samrichca @b-f-chan just one question. Do we need to display this root level "Song"?
That will be shown in the titlebar. 

![image.png](https://images.zenhubusercontent.com/5a5f5bcd4b5806bc2bc31a86/24d694cb-e908-426d-a5b4-9d4473677414)

![image.png](https://images.zenhubusercontent.com/5a5f5bcd4b5806bc2bc31a86/f9592701-00a4-468b-92f5-7e82baf0beaa)

**Comment by samrichca** [comment]
@kcullion nope we don't! i'll remove it as we go, if that's ok.

**Comment by kcullion** [comment]
sounds good @samrichca thanks!

**Comment by samrichca** [comment]
the file/URL structure here is out of date, see: https://github.com/overture-stack/roadmap/issues/49

**Comment by b-f-chan** [comment]
Reviewed the file/URL structure with @samrichca and it looks reasonable to me.  Agreed with @samrichca scope of this ticket can be considered completed, we have arrived at a workable structure.

Although the netlify links provided above are outdated, we will test the final structure in roadmap#49.
